cli/src/main.rs

```diff
@@ -6,25 +6,46 @@
 
 mod commands;
 use clap::Parser as ClapParser;
+use clap::{crate_authors, crate_version, ColorChoice, Parser};
+use subxt_codegen::utils::Uri;
 
 /// Subxt utilities for interacting with Substrate based nodes.
 #[derive(Debug, ClapParser)]
-enum Command {
-    Metadata(commands::metadata::Opts),
-    Codegen(commands::codegen::Opts),
-    Compatibility(commands::compatibility::Opts),
-    Version(commands::version::Opts),
+pub enum SubCommand {
+    Metadata(commands::metadata::MetadataOpts),
+    Codegen(commands::codegen::CodegenOpts),
+    Compatibility(commands::compatibility::CompatOpts),
+    Version(commands::version::Version),
+}
+
+#[derive(Debug, Parser)]
+#[clap(version = crate_version!(), author = crate_authors!(), color=ColorChoice::Always)]
+pub struct CliOpts {
+    #[clap(subcommand)]
+    pub subcmd: SubCommand,
+
+    /// The url of the substrate node connect to
+    #[clap(
+        name = "url",
+        long,
+        value_parser,
+        default_value = "http://localhost:9933",
+        env = "SUBXT_URL"
+    )]
+    url: Uri,
 }
 
 #[tokio::main]
 async fn main() -> color_eyre::Result<()> {
     color_eyre::install()?;
-    let args = Command::parse();
+    let opts = CliOpts::parse();
 
-    match args {
-        Command::Metadata(opts) => commands::metadata::run(opts).await,
-        Command::Codegen(opts) => commands::codegen::run(opts).await,
-        Command::Compatibility(opts) => commands::compatibility::run(opts).await,
-        Command::Version(opts) => commands::version::run(opts),
+    match &opts.subcmd {
+        SubCommand::Metadata(cmd_opts) => commands::metadata::run(&opts, &cmd_opts).await,
```

Comment:
Also you end up with this sort of thing which I find a bit weird; `commands::metadata::run(&opts, &cmd_opts)` is now taking in _all_ of the opts, but it's also taking in part of the same thing again, the SubCommand opts.

Reply:
Yes. I think this is a trade off and also you sharing some global flags (`--json` is another typical one) with all the commands. Sure there are cases where it is non-sense (ie `subxst --url <bla> version`) but that would nonetheless work and not get in the way of the user.

Reply:
Are you concerned about the amount of extra data we pass ?

Reply:
Oh nothing like that. For me, I think that in general, moving out the url from each command reduces the clarity and maintainability of the code. It's very simple to understand at present that for each command you'd like to add/work on, you can specify exactly the options that it wants and you can see all options specified in the same file as the command logic. 

I don't want us to go down a path that will result in "global" options eg url (because _most_ commands use them) and then the extra muddle in the code that this introduces. For me it's just not worth it. What I am much more in favour of is things like helper structs which encompass "special URL command logic" that we can use to easily gain that URL logic in any command we want to use it for.

Reply:
I understand your logic but I respectfully disagree for the following reasons:
- code and doc duplication: We can see the drawback currently with the doc of the codegen's url  flag that mentions the metadata command, it is probably the remain of a copy/paste
- it suggests that the url for the `metadata` command does not have the same meaning than the url of the `codegen` command. I would be in favor of keeping `--url` local if the url would be pointing to different things but for subxt, they point to **the** endpoint url AFAIK

I don't know all the plans for `subxt` and if it is not aimed at becoming primarily a tool to send transactions or query a node reached via **the** `url` of its endpoint, then I would agree that keeping things local and dealing with the duplication'costs could be worth it. But I think `subxt` is primarily targetted at connecting to a node, even if some utility commands may be part of it and not require a url.

If you plan on adding later for instance a `--json` flag so all outputs are _automation friendly_, it is MUCH easier to have that also global that spreading dupllication is most or all commands. That also means you'd need to **add** this `--json` flag (or `--url`) in every single upcoming new command, which is quite some hassle for a small gain.